### PR TITLE
Implement Houdini worklets

### LINKED ISSUE #16206: Implement CSS Houdini Worklets
Draft spec: https://drafts.css-houdini.org/worklets/
Houdini CSS issue for servo: #13557 

cc @tschneidereit @metajack @jdm @SimonSapin

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR implements the current draft Houdini Worklets specification (https://drafts.css-houdini.org/worklets/).

The implementation is intended to provide a responsive environment for worklet execution, and in particular to ensure that the primary worklet executor does not garbage collect, and does not block loading module code. The implementation does this by providing a thread pool, and performing GC and module loading in a backup thread, not in the primary thread.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #16206 
- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16814)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @tschneidereit @metajack

**Comment by metajack** [comment]
cc @bfgeek

**Comment by asajeffrey** [comment]
Some things that should probably be fixed before this lands:
 * the methods in `globalscope` which use downcasting are missing some `WorkletGlobalScope` cases.
 * threads should be named.
 * add time reporting code.
 * think about how to configure SM for worklet threads (in particular how to discourage GC).

The code is ready for review as-is though.

**Comment by asajeffrey** [comment]
Hi @bfgeek! This is implementation of worklets in Servo, using a thread pool to ensure worklet execution isn't blocked by GC or by module loading. Mostly the spec was pretty straightforward to implement, you'll be pleased to hear!

**Comment by asajeffrey** [comment]
The spec issues raised by implementing worklets (and starting on paint worklets):

* Speculative evalution (https://github.com/w3c/css-houdini-drafts/issues/406)
* Fetching and invoking a worklet script which reports an exception (https://github.com/w3c/css-houdini-drafts/issues/407)

**Comment by bfgeek** [comment]
cc/ @nhiroki who works on our implementation. 

That's great to see! Thanks for filing spec issues.

File: components/script/dom/webidls/Window.webidl [hunk 1]
```diff
@@ -201,3 +201,10 @@ partial interface Window {
    readonly attribute TestRunner testRunner;
    //readonly attribute EventSender eventSender;
 };
+
+// https://drafts.css-houdini.org/worklets/#examples
+partial interface Window {
+   [Pref="dom.worklet.testing.enabled", SameObject] readonly attribute Worklet testWorklet;
+   [Pref="dom.worklet.testing.enabled"] DOMString? testWorkletLookup(DOMString key);
```
Comment (jdm) [inline components/script/dom/webidls/Window.webidl]: What if we made a separate DOM interface with a constructor so we don't need to include test-only fields in Window?

File: components/script/dom/workletglobalscope.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+/// Resources required by workletglobalscopes
+#[derive(Clone)]
+pub struct WorkletGlobalScopeInit {
+    /// Chan to a resource thread
+    pub resource_threads: ResourceThreads,
+    /// Chan to the memory profiler
+    pub mem_profiler_chan: mem::ProfilerChan,
+    /// Chan to the time profiler
+    pub time_profiler_chan: time::ProfilerChan,
+    /// Chan to devtools
```
Comment (jdm) [inline components/script/dom/workletglobalscope.rs]: Let's use channel instead of chan in these comments.

File: components/script/dom/worklet.rs [hunk 3]
```diff
@@ -0,0 +1,621 @@
...
+use std::sync::atomic::AtomicIsize;
+use std::sync::atomic::Ordering;
+use std::sync::mpsc;
+use std::sync::mpsc::Receiver;
+use std::sync::mpsc::Sender;
+use std::thread;
+use style::thread_state;
+use swapper::Swapper;
+use swapper::swapper;
+use uuid::Uuid;
```
Comment (jdm) [inline components/script/dom/worklet.rs]: Any reason not to collapse some of these import statements from the same modules?

File: components/script/dom/worklet.rs [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+
+    #[allow(unsafe_code)]
+    #[allow(unrooted_must_root)]
+    pub fn new(window: &Window, global_type: WorkletGlobalScopeType) -> Root<Worklet> {
+        debug!("Creating worklet {:?}.", global_type);
+        let worklet = box Worklet::new_inherited(window, global_type);
+        let global = window.upcast::<GlobalScope>();
+        unsafe {
+            WorkletBinding::Wrap(global.get_cx(), global, worklet)
+        }
```
Comment (jdm) [inline components/script/dom/worklet.rs]: We should use `reflect_dom_object` instead.

File: components/script/dom/worklet.rs [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+    }
+
+    pub fn worklet_id(&self) -> WorkletId {
+        self.worklet_id
+    }
+
+    #[allow(dead_code)]
```
Comment (jdm) [inline components/script/dom/worklet.rs]: ?

File: components/script/dom/worklet.rs [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+        // Step 5.
+        promise
+    }
+}
+
+/// A guid for worklets.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, JSTraceable)]
+pub struct WorkletId(Uuid);
+
+known_heap_size!(0, WorkletId);
```
Comment (jdm) [inline components/script/dom/worklet.rs]: I would prefer using `#[ignore_heap_size_of = "Can't measure uuid"]` instead.

File: components/script/dom/worklet.rs [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+    fn new() -> PendingTasksStruct {
+        PendingTasksStruct(Arc::new(AtomicIsize::new(WORKLET_THREAD_POOL_SIZE as isize)))
+    }
+
+    fn set_counter_to(&self, value: isize) -> isize {
+        self.0.swap(value, Ordering::AcqRel)
+    }
+
+    fn decrement_counter_by(&self, offset: isize) -> isize {
+        self.0.fetch_sub(offset, Ordering::AcqRel)
```
Comment (jdm) [inline components/script/dom/worklet.rs]: Why AcqRel instead of SeqCst?

File: components/script/dom/worklet.rs [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+            let old_counter = pending_tasks_struct.set_counter_to(-1);
+            if old_counter > 0 {
+                self.run_in_script_thread(promise.reject_runnable(Error::Abort));
+            }
+        } else {
+            // Step 5.
+            debug!("Finished adding script.");
+            let old_counter = pending_tasks_struct.decrement_counter_by(1);
+            if old_counter == 1 {
+                // TODO: trigger a reflow?
```
Comment (jdm) [inline components/script/dom/worklet.rs]: This happens automatically when events run in the script thread. Did you mean something more specific?

File: components/script/dom/worklet.rs [hunk 9]
```diff
@@ -0,0 +1,621 @@
...
+            hash_map::Entry::Vacant(entry) => {
+                let result = global_type.new(&self.runtime, pipeline_id, base_url, &self.global_init);
+                entry.insert(JS::from_ref(&*result));
+                result
+            },
+        }
+    }
+
+    /// Fetch and invoke a worklet script.
+    ///g https://drafts.css-houdini.org/worklets/#fetch-and-invoke-a-worklet-script
```
Comment (jdm) [inline components/script/dom/worklet.rs]: nit: g

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
As usual, I found these changes to be a pleasant and straightforward read. Good job!

File: components/script/dom/worklet.rs [hunk 10]
```diff
@@ -0,0 +1,621 @@
...
+        // TODO: Fetch a module graph, not just a single script.
+        // TODO: Fetch the script asynchronously?
+        // TODO: Caching.
+        let resource_fetcher = self.global_init.resource_threads.sender();
+        let request = RequestInit {
+            url: script_url,
+            type_: RequestType::Script,
+            destination: Destination::Script,
+            credentials_mode: credentials.into(),
+            .. RequestInit::default()
```
Comment (jdm) [inline components/script/dom/worklet.rs]: We should be using CORS mode here, per https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-single-module-script. We'll need to include a real origin, too.

File: components/script/dom/webidls/Window.webidl (hunk 1 above)
Comment (asajeffrey) [inline components/script/dom/webidls/Window.webidl]: Will do.

File: components/script/dom/workletglobalscope.rs (hunk 2 above)
Comment (asajeffrey) [inline components/script/dom/workletglobalscope.rs]: Will do.

File: components/script/dom/worklet.rs (hunk 3 above)
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: Minimizing the chance of annoying merge conflicts.

File: components/script/dom/worklet.rs (hunk 4 above)
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: I don't think you can use `reflect_dom_object` to create globals can you?

File: components/script/dom/worklet.rs (hunk 5 above)
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: This is dead code, just in for future-proofing. I can get rid of it.

File: components/script/dom/worklet.rs (hunk 6 above)
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: Will do.

File: components/script/dom/worklet.rs (hunk 7 above)
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: Using the weakest ordering that does the job. In this case it doesn't matter a huge amount, this isn't hot code.

File: components/script/dom/worklet.rs (hunk 9 above)
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: Oops.

File: components/script/dom/worklet.rs (hunk 10 above)
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: CORS mode isn't a problem, but I'm not sure which origin to use here. One snag is that this is being done in a different thread, so we don't have access to the `MutableOrigin` of the window.

File: components/script/dom/worklet.rs (hunk 8 above)
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: At this point, we have loaded a worklet, but not actually run it (e.g. to get a paint worklet to do its painting).

**Review by asajeffrey — COMMENTED** [review]
Thanks for the review!

File: components/script/dom/worklet.rs (hunk 4 above)
Comment (jdm) [inline components/script/dom/worklet.rs]: WorkletBinding isn't a global.

File: components/script/dom/worklet.rs (hunk 10 above)
Comment (jdm) [inline components/script/dom/worklet.rs]: Send an ImmutableOrigin from the window as part of the initiation?

File: components/script/dom/worklet.rs (hunk 4 above)
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: Oh good point, we're in `Worklet` not `WorkletGlobalScope`.

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
Squash and r=me.

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface Worklet to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:73:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:3
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:13:1
```
We can either hide the Worklet interface behind an off-by-default pref or update the test to include the interface in the expected list.

**Comment by asajeffrey** [comment]
I added `Worklet` to the interfaces test. @bors-servo r=jdm

**Pull request merged**